File: frontend/src/types/analytics.ts
// types/analytics.ts
export interface UserAnalytics {
  userId: string;
  totalPapers: number;
  publishedPapers: number;
  draftPapers: number;
  inProgressPapers: number;
  totalWords: number;
  avgProgress: number;
  totalCollaborators: number;
  researchAreas: number;
  avgCompletionTime: number;
  productivityScore: number;
  lastUpdated: Date;
}

export interface PaperAnalytics {
  paperId: string;
  wordCount: number;
  readingTime: number;
  estimatedCompletionTime: number;
  collaboratorCount: number;
  sectionProgress: SectionAnalytics[];
  timeDistribution: TimeDistribution[];
  citationCount: number;
  revisionCount: number;
  createdAt: Date;
  lastModified: Date;
}

export interface SectionAnalytics {
  sectionId: string;
  title: string;
  wordCount: number;
  status: string;
  timeSpent: number;
  lastModified: Date;
  revisions: number;
}

export interface TimeDistribution {
  date: string;
  wordsWritten: number;
  timeSpent: number;
  sessionsCount: number;
}

export interface ProductivityMetrics {
  daily: DailyMetric[];
  weekly: WeeklyMetric[];
  monthly: MonthlyMetric[];
}

export interface DailyMetric {
  date: string;
  wordsWritten: number;
  timeSpent: number;
  papersWorkedOn: number;
  sessionsCount: number;
  focusScore: number;
}

export interface WeeklyMetric {
  week: string;
  totalWords: number;
  totalTime: number;
  averageWordsPerDay: number;
  mostProductiveDay: string;
  papersProgress: number;
}

export interface MonthlyMetric {
  month: string;
  papersCompleted: number;
  totalWords: number;
  averageWordsPerWeek: number;
  collaborations: number;
  milestonesReached: number;
}

export interface CollaborationAnalytics {
  collaboratorId: string;
  name: string;
  email: string;
  sharedPapers: number;
  totalContributions: number;
  avgResponseTime: number;
  collaborationStarted: Date;
  lastActivity: Date;
  contributionTypes: ContributionType[];
}

export interface ContributionType {
  type: 'writing' | 'editing' | 'review' | 'methodology' | 'analysis';
  count: number;
  percentage: number;
}

export interface ResearchTrend {
  area: string;
  paperCount: number;
  wordCount: number;
  averageProgress: number;
  timeSpent: number;
  collaborators: string[];
  publications: number;
  citations: number;
}

export interface WritingPattern {
  preferredTime: {
    hour: number;
    dayOfWeek: number;
    productivity: number;
  };
  sessionLength: {
    average: number;
    optimal: number;
    distribution: number[];
  };
  writingVelocity: {
    wordsPerHour: number;
    wordsPerSession: number;
    consistency: number;
  };
}

export interface AnalyticsFilter {
  dateRange: {
    start: Date;
    end: Date;
  };
  paperIds?: string[];
  collaborators?: string[];
  researchAreas?: string[];
  status?: string[];
}

export interface AnalyticsExport {
  format: 'csv' | 'json' | 'pdf';
  data: any;
  generatedAt: Date;
  filters: AnalyticsFilter;
}

export interface Insight {
  id: string;
  type: 'productivity' | 'collaboration' | 'writing' | 'research';
  title: string;
  description: string;
  severity: 'info' | 'warning' | 'critical';
  actionable: boolean;
  suggestions: string[];
  dataPoints: any[];
  generatedAt: Date;
}

export interface ComparisonMetric {
  label: string;
  userValue: number;
  benchmarkValue: number;
  percentile: number;
  trend: 'improving' | 'declining' | 'stable';
}

export interface ResearchGoal {
  id: string;
  title: string;
  target: number;
  current: number;
  unit: string;
  deadline: Date;
  type: 'words' | 'papers' | 'time' | 'collaborations';
  priority: 'high' | 'medium' | 'low';
  achieved: boolean;
}

export type AnalyticsTimeframe = 'day' | 'week' | 'month' | 'quarter' | 'year' | 'all';
export type MetricType = 'papers' | 'words' | 'time' | 'collaborations' | 'productivity';
export type ChartType = 'line' | 'bar' | 'pie' | 'area' | 'scatter';

File: backend/app/schemas/analytics.py
"""
Analytics Schemas (app/schemas/analytics.py)
"""
from pydantic import BaseModel, Field
from typing import List, Optional, Dict, Any
from datetime import datetime


class AnalyticsOverview(BaseModel):
    """Analytics overview schema matching service output"""
    user_id: str
    timeframe: str
    total_papers: int = 0
    published_papers: int = 0
    draft_papers: int = 0
    in_progress_papers: int = 0
    total_words: int = 0
    avg_progress: float = 0.0
    total_collaborators: int = 0
    research_areas: int = 0
    avg_completion_time: float = 0.0
    productivity_score: int = 0
    last_updated: str

    # Additional nested data that service might return
    collaboration: Optional[Dict[str, Any]] = Field(default_factory=lambda: {"total_collaborators": 0})

    class Config:
        from_attributes = True


class TrendDataPoint(BaseModel):
    """Single data point for trend analysis"""
    date: str
    value: float
    label: Optional[str] = None


class TrendAnalysis(BaseModel):
    """Trend analysis schema"""
    period: str  # "day", "week", "month", "quarter", "year"
    metric: str
    data_points: List[TrendDataPoint] = []
    total_change: float = 0.0
    percentage_change: float = 0.0
    trend_direction: str = "stable"  # "up", "down", "stable"

    class Config:
        from_attributes = True


class SectionProgress(BaseModel):
    """Individual section progress"""
    section_id: str
    title: str
    word_count: int
    status: str
    progress: float
    last_modified: str


class PaperAnalytics(BaseModel):
    """Detailed paper analytics"""
    paper_id: str
    word_count: int
    reading_time: int
    estimated_completion_time: float
    collaborator_count: int
    section_progress: List[SectionProgress]
    revision_count: int
    created_at: str
    last_modified: str

    class Config:
        from_attributes = True


class ProductivityDataPoint(BaseModel):
    """Daily productivity metrics"""
    date: str
    words_written: int
    papers_worked_on: int
    sessions_count: int
    focus_score: int
    time_spent: Optional[float] = 0.0


class ProductivityMetrics(BaseModel):
    """Productivity metrics over time"""
    daily: List[ProductivityDataPoint]
    start_date: str
    end_date: str

    class Config:
        from_attributes = True


class CollaborationAnalytics(BaseModel):
    """Collaboration analytics for a collaborator"""
    collaborator_name: str
    shared_papers: int
    total_words: int
    research_areas: List[str]
    collaboration_strength: int

    class Config:
        from_attributes = True


class ResearchTrend(BaseModel):
    """Research trend by area"""
    area: str
    paper_count: int
    word_count: int
    average_progress: float
    collaborators: List[str]
    publications: int

    class Config:
        from_attributes = True


class WritingPattern(BaseModel):
    """Writing patterns and habits"""
    average_words_per_paper: float
    average_sections_per_paper: float
    most_productive_status: str
    consistency_score: float
    total_writing_days: int

    class Config:
        from_attributes = True


class Insight(BaseModel):
    """AI-generated insight"""
    type: str
    title: str
    description: str
    severity: str
    actionable: bool
    suggestions: Optional[List[str]] = None

    class Config:
        from_attributes = True


class AnalyticsResponse(BaseModel):
    """Complete analytics response"""
    overview: AnalyticsOverview
    trends: Optional[List[TrendAnalysis]] = None
    insights: Optional[List[Insight]] = None

    class Config:
        from_attributes = True